stub_output_writer now dynamically reallocates memory

viniul opened the pull request:
Memory allocation for stub_output_writer is now dynamic:
gifData will be reallocated in base 2 steps every time
we need more space.

viniul committed:
stub_output_writer now dynamically reallocates memory

Memory allocation for stub_output_writer is now dynamic:
gifData will be reallocated in base 2 steps every time
we need more space.

viniul commented:
@bshastry Please check in particular for memory leaks/uafs/double frees.

bshastry commented on projects/giflib/egif_fuzz_common.cc:
```diff
@@ -1,26 +1,36 @@
+	if (gud->allocatedSize < (gud->gifLen + len))
+	{
+		// Reallocate gifFileType
+		int newSize = gud->allocatedSize * 2;
```
If `newSize` is still less than `gud->gifLen + len`, buffer over runs.

viniul replied:
Whops, really good catch! What is the best way to deal with that:
std::max(newSize, gud->gifLen + len);?

bshastry replied:
perhaps `int newSize = (gud->gifLen + len) * 2;`?

viniul replied:
sounds good

viniul replied:
Fixed :)

bshastry commented on projects/giflib/egif_fuzz_common.cc:
```diff
@@ -1,26 +1,36 @@
+		if (gud->gifData == NULL)
+		{
+			gud->gifData = oldGud;
+			return 0;
```
Should we add a debug print here under the FUZZING_MODE_UNSAFE_FOR_PRODUCTION macro so we know how often, if at all, this happens?

viniul replied:
Yup - my intuition here is that this should only happen when the target is OOM, in which case the fuzzer should complain anyway, correct?

bshastry replied:
Good point. To be safe, we could add `assert(0)` here so we know if some input gets us here before OOM.

viniul replied:
Will do:)

viniul replied:
Fixed

bshastry commented on projects/giflib/egif_fuzz_common.cc:
```diff
@@ -1,26 +1,36 @@
 // RGB to GIF converter
-static
-bool rgb_to_gif(const uint8_t* data, size_t size)
+static bool rgb_to_gif(const uint8_t *data, size_t size)
```
Ignoring the formatting changes since I feel they were due to an IDE.

viniul replied:
Yup - sorry for polluting the commit.

bshastry replied:
No worries, just wanted to mention I didn't review these lines ;)

Commit:
Fix buffer overrun in stub_output_writer when reallocating memory

viniul committed:
Throw assertion if memory reallocation in egif_target fails

egif_target uses dynamic memory reallocation to assure
that stub_output_writer never has a buffer overrun
in stub_output_writer. We throw an assert if
this memory reallocation fails.

bshastry reviewed (approved):
LGTM
